system/jlib/jfile.cpp

```diff
@@ -4441,6 +4441,76 @@ IFileIOStream * createBufferedIOStreamFromFile(const char *fileNameWithPath, IFO
     return createBufferedIOStream(iFileIO, bufsize);
 }
 
+IFileIOStream *createProgressIFileIOStream(IFileIOStream *iFileIOStream, offset_t totalSize, const char *msg, unsigned periodSecs)
+{
+    class CProgressIFileIOStream : public CSimpleInterfaceOf<IFileIOStream>
+    {
+        Linked<IFileIOStream> iFileIOStream;
+        offset_t totalSize = 0;
+        StringAttr msg;
+        cycle_t timePeriodCycles = 0;
+        cycle_t nextLogCycles = 0;
+        void log(double pct)
+        {
+            PROGLOG("%s - %.2f%% complete", msg.get(), pct);
+        }
+    public:
+        CProgressIFileIOStream(IFileIOStream *_iFileIOStream, offset_t _totalSize, const char *_msg, unsigned periodSecs)
+            : iFileIOStream(_iFileIOStream), totalSize(_totalSize), msg(_msg)
+        {
+            timePeriodCycles = queryOneSecCycles() * periodSecs;
+            nextLogCycles = get_cycles_now() + timePeriodCycles;
+        }
+        ~CProgressIFileIOStream()
+        {
+            offset_t pos = iFileIOStream->tell();
+            if (pos == totalSize)
+                log(100.0);
+        }
+        // implements ISimpleReadStream
+        virtual size32_t read(size32_t max_len, void * data) override
+        {
+            cycle_t nowCycles = get_cycles_now();
+            if (nowCycles >= nextLogCycles)
```

Comment:
Trivial: This could theoretically wrap - although would probably need an uptime of 100 years...

Reply:
yes, I did realize, but didn't think worth worrying about too much, and there are many other places that would similarly fall foul if this was so, including all uses of CCycleTimer::elapsed* functions.

Reply:
> > @jakesmith I have merged that other PR. Do you want me to merge this as-is, or experiment with using that class?
> 
> I'll incorporate the new helper class in this PR.

@ghalliday - have now incorporated PeriodicTimer